File: src/pipeline/reframe.ts
import path from "node:path";
import { fileURLToPath } from "node:url";
import { config } from "../config.js";
import { run } from "../utils/exec.js";
import { log } from "../utils/logger.js";

export type LayoutMode = "auto" | "single" | "split" | "center";

export interface ReframePlan {
  /** "center" = plain center crop, "single" = one centered subject,
   *  "split" = two stacked subjects. */
  layout: "center" | "single" | "split";
  /** horizontal center as a fraction 0..1 (single layout) */
  center?: number;
  /** two horizontal centers as fractions 0..1 (split layout, top then bottom) */
  centers?: [number, number];
}

const HELPER = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "../../scripts/reframe.py"
);

/** Parse a --focus value into a 0..1 fraction, or null if not provided. */
export function parseFocus(value: string | undefined): number | null {
  if (value == null || value === "") return null;
  const v = value.toLowerCase();
  if (v === "center" || v === "centre") return 0.5;
  if (v === "left") return 0.25;
  if (v === "right") return 0.75;
  const n = parseFloat(v);
  if (Number.isFinite(n) && n >= 0 && n <= 1) return n;
  return null;
}

/**
 * Decide how to crop a clip. Tries the OpenCV helper for face-aware framing;
 * always falls back to a center crop if anything goes wrong.
 */
export async function planReframe(
  video: string,
  startSec: number,
  durationSec: number,
  mode: LayoutMode,
  focus: number | null
): Promise<ReframePlan> {
  // explicit manual override always wins
  if (focus != null) return { layout: "single", center: focus };
  if (mode === "center") return { layout: "center" };

  try {
    const { stdout } = await run(config.python.bin, [
      HELPER,
      path.resolve(video),
      startSec.toFixed(3),
      durationSec.toFixed(3),
      "12",
      mode,
    ]);
    const data = JSON.parse(stdout.trim()) as {
      layout?: string;
      center?: number;
      centers?: [number, number];
      error?: string;
    };

    if (data.error) {
      log.warn(`reframe fell back to center crop (${data.error})`);
      return { layout: "center" };
    }
    if (data.layout === "split" && data.centers) {
      return { layout: "split", centers: data.centers };
    }
    if (data.layout === "single" && typeof data.center === "number") {
      return { layout: "single", center: data.center };
    }
    return { layout: "center" };
  } catch (err) {
    log.warn(
      `reframe helper unavailable, using center crop (${err instanceof Error ? err.message : String(err)})`
    );
    return { layout: "center" };
  }
}

File: scripts/reframe.py
#!/usr/bin/env python3
"""
Auto-reframe helper for auto-clipper.

Samples frames from a time window of a video, detects faces with OpenCV's
Haar cascade, and decides how to crop the clip to vertical:

  - "single": one subject  -> returns one horizontal center (fraction 0..1)
  - "split" : two subjects -> returns two centers (top, bottom)
  - "center": no/unclear faces -> caller should center-crop

Output: a single JSON line on stdout, e.g.
  {"width":1280,"height":720,"layout":"split","centers":[0.31,0.72]}

Designed to fail soft: if OpenCV is missing or anything goes wrong, it prints
{"layout":"center"} so the Node side falls back to a plain center crop.
"""
import sys
import json


def emit(obj):
    sys.stdout.write(json.dumps(obj))
    sys.stdout.flush()


def main():
    if len(sys.argv) < 4:
        emit({"layout": "center", "error": "usage: reframe.py <video> <start> <duration> [samples] [mode]"})
        return

    video = sys.argv[1]
    start = float(sys.argv[2])
    duration = float(sys.argv[3])
    samples = int(sys.argv[4]) if len(sys.argv) > 4 else 12
    mode = sys.argv[5] if len(sys.argv) > 5 else "auto"

    try:
        import cv2
        import numpy as np
    except Exception as e:  # OpenCV/numpy not installed
        emit({"layout": "center", "error": f"import failed: {e}"})
        return

    cap = cv2.VideoCapture(video)
    if not cap.isOpened():
        emit({"layout": "center", "error": "cannot open video"})
        return

    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)) or 0
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)) or 0
    cascade = cv2.CascadeClassifier(
        cv2.data.haarcascades + "haarcascade_frontalface_default.xml"
    )

    centers = []  # (x_center_px, area)
    min_face = max(24, int(height * 0.06))
    for i in range(samples):
        t = start + duration * (i + 0.5) / max(1, samples)
        cap.set(cv2.CAP_PROP_POS_MSEC, t * 1000.0)
        ok, frame = cap.read()
        if not ok:
            continue
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        faces = cascade.detectMultiScale(
            gray, scaleFactor=1.1, minNeighbors=5, minSize=(min_face, min_face)
        )
        for (x, y, w, h) in faces:
            centers.append((x + w / 2.0, float(w * h)))

    cap.release()

    if not centers or width == 0:
        emit({"width": width, "height": height, "layout": "center"})
        return

    xs = np.array([c[0] for c in centers], dtype=float)
    areas = np.array([c[1] for c in centers], dtype=float)

    def single():
        c = float(np.average(xs, weights=areas))
        return {"width": width, "height": height, "layout": "single", "center": c / width}

    def try_split():
        order = np.argsort(xs)
        xs_sorted = xs[order]
        if len(xs_sorted) < 4:
            return None
        gaps = np.diff(xs_sorted)
        gi = int(np.argmax(gaps))
        if gaps[gi] < width * 0.18:
            return None
        left = xs_sorted[: gi + 1]
        right = xs_sorted[gi + 1 :]
        if len(left) < 2 or len(right) < 2:
            return None
        c1 = float(np.median(left)) / width
        c2 = float(np.median(right)) / width
        lo, hi = sorted((c1, c2))
        return {"width": width, "height": height, "layout": "split", "centers": [lo, hi]}

    if mode == "single":
        emit(single())
        return
    if mode == "split":
        emit(try_split() or single())
        return

    # auto
    emit(try_split() or single())


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        emit({"layout": "center", "error": str(e)})
